gobblin-utility/src/main/java/gobblin/util/fileSystem/PathAlterationObserver.java

```diff
@@ -0,0 +1,352 @@
+
+
+public class PathAlterationObserver {
+
+  private final List<PathAlterationListener> listeners = new CopyOnWriteArrayList<>();
+  private final PathEntry rootEntry;
+  private final PathFilter pathFilter;
+  private final Comparator<Path> comparator;
+  private final FileSystem fs;
+
+  private final Path[] EMPTY_PATH_ARRAY = new Path[0];
+
+  /**
+   * Final processing.
+   *
+   * @throws Exception if an error occurs
+   */
+  public void destroy()
+      throws Exception {
+  }
+
+  /**
+   * Construct an observer for the specified directory.
+   *
+   * @param directoryName the name of the directory to observe
+   */
+  public PathAlterationObserver(final String directoryName)
+      throws IOException {
+    this(new Path(directoryName));
+  }
+
+  /**
+   * Construct an observer for the specified directory and file filter.
+   *
+   * @param directoryName the name of the directory to observe
+   * @param pathFilter The file filter or null if none
+   */
+  public PathAlterationObserver(final String directoryName, final PathFilter pathFilter)
+      throws IOException {
+    this(new Path(directoryName), pathFilter);
+  }
+
+  /**
+   * Construct an observer for the specified directory, file filter and
+   * file comparator.
+   *
+   * @param directoryName the name of the directory to observe
+   * @param pathFilter The file filter or null if none
+   * @param caseSensitivity  what case sensitivity to use comparing file names, null means system sensitive
+   */
+  public PathAlterationObserver(final String directoryName, final PathFilter pathFilter, final IOCase caseSensitivity)
+      throws IOException {
+    this(new Path(directoryName), pathFilter, caseSensitivity);
+  }
+
+  /**
+   * Construct an observer for the specified directory.
+   *
+   * @param directory the directory to observe
+   */
+  public PathAlterationObserver(final Path directory)
+      throws IOException {
+    this(directory, null);
+  }
+
+  /**
+   * Construct an observer for the specified directory and file filter.
+   *
+   * @param directory the directory to observe
+   * @param pathFilter The file filter or null if none
+   */
+  public PathAlterationObserver(final Path directory, final PathFilter pathFilter)
+      throws IOException {
+    this(directory, pathFilter, null);
+  }
+
+  /**
+   * Construct an observer for the specified directory, file filter and
+   * file comparator.
+   *
+   * @param directory the directory to observe
+   * @param fileFilter The file filter or null if none
+   * @param caseSensitivity  what case sensitivity to use comparing file names, null means system sensitive
+   */
+  public PathAlterationObserver(final Path directory, final PathFilter fileFilter, final IOCase caseSensitivity)
+      throws IOException {
+    this(new PathEntry(directory), fileFilter, caseSensitivity);
+  }
+
+  public PathAlterationObserver(final PathEntry rootEntry, final PathFilter pathFilter, final IOCase caseSensitivity)
```

Comment:
Let's not worry about case sensitivity. In Hadoop everything is case sensitive, it will reduce the necessary code.